include/swift/IDE/CodeCompletion.h

```diff
@@ -630,21 +630,287 @@ enum class CodeCompletionDiagnosticSeverity : uint8_t {
   MAX_VALUE = Note
 };
 
-/// A single code completion result.
-class CodeCompletionResult {
-  friend class CodeCompletionResultBuilder;
+/// Reasons why a code completion item might not be recommended in a certain
+/// context.
+/// This enum is split into two subsets: \c ContextFreeNotRecommendedReason and
+/// \c ContextualNotRecommendedReason. When adding a case to this enum also add
+/// it to either of those.
+/// Context-free diagnostics are independent of the context they are used in.
+/// The not-recommended reason can thus be cached as part of a context free
+/// code completion result.
+/// Contextual not recommended reasons depend on the context they are used in.
+/// E.g. \c InvalidAsyncContext depends on whether the usage context is async or
+/// not.
+enum class NotRecommendedReason : uint8_t {
+  None = 0,                    // both contextual and context-free
+  RedundantImport,             // contextual
+  RedundantImportIndirect,     // contextual
+  Deprecated,                  // context-free
+  SoftDeprecated,              // context-free
+  InvalidAsyncContext,         // contextual
+  CrossActorReference,         // contextual
+  VariableUsedInOwnDefinition, // contextual
+
+  MAX_VALUE = VariableUsedInOwnDefinition
+};
+
+/// TODO: We consider deprecation warnings as context free although they don't
+/// produce deprecation warnings insdide context that have the same or a
+/// stronger deprecation attribute.
+/// E.g.
+/// In SDK:
+/// \code
+/// @available(iOS, deprecated: 12.0)
+/// func deprecatedFunc()
+/// \endcode
+///
+/// User code:
+/// \code
+/// @available(iOS, deprecated: 12.0)
+/// func insiderUserDeprecatedContext() {
+///   #^COMPLETE^#
+/// }
+/// \endcode
+/// suggests \c deprecatedFunc as deprecated although it doesn't produce a
+/// diagnostic during compilation. But this allows us to show deprecation
+/// warnings for cached results.
+enum class ContextFreeNotRecommendedReason : uint8_t {
+  None = 0,
+  Deprecated,
+  SoftDeprecated,
+  MAX_VALUE = SoftDeprecated
+};
+
+enum class ContextualNotRecommendedReason : uint8_t {
+  None = 0,
+  RedundantImport,
+  RedundantImportIndirect,
+  InvalidAsyncContext,
+  CrossActorReference,
+  VariableUsedInOwnDefinition,
+  MAX_VALUE = VariableUsedInOwnDefinition
+};
+
+enum class CodeCompletionResultKind : uint8_t {
+  Declaration,
+  Keyword,
+  Pattern,
+  Literal,
+  BuiltinOperator,
+
+  MAX_VALUE = BuiltinOperator
+};
+
+/// The parts of a \c CodeCompletionResult that are not dependent on the context
+/// it appears in and can thus be cached.
+class ContextFreeCodeCompletionResult {
+  CodeCompletionResultKind Kind : 3;
+  static_assert(int(CodeCompletionResultKind::MAX_VALUE) < 1 << 3, "");
+
+  /// Opaque kind that is interpreted differently depending on
```

Comment:
Nit: Incomplete comment? Probably:
```suggestion
  /// Opaque kind that is interpreted differently depending on 'Kind'.
```

Reply:
Thanks. That comment pre-dated the use of a union here. I just removed it.